### ISSUE TITLE
Production build times massively increased after use of broccoli funnel with TinyMCE

### ISSUE THREAD

**Issue by markstuart** [ember-cli/ember-cli#5562]
As part of an application I am involved in building, we have added TinyMCE as a rich text editor using `broccoli-funnel` in the way advocated by the ember-cli guidelines. Since the addition of this library, our `ember build --prod` build times have skyrocketed.

I have investigated the other similar issues around build times with ember-cli, and tried to follow your performance document to improve things, but without success at this point.

If someone could take a look at the example project I have set up and verify my findings or point me in the right direction to fixing this issue that would be great :)

Example project (including my system and node evnironment info) is [here](https://github.com/markstuart/ember-build-prod-test)

My main finding at this stage (also in the readme for that project) is that the build process appears to just hang for a long period of time straight after `broccoli-merge-trees:TreeMerger (allTrees) build:` is output to the console (when running `DEBUG=broccoli*,ember-cli*` with the build command)

**Comment by stefanpenner** [comment]
The performance document provides information you should provide when reporting a performance issue. Please update accordingly.

**Comment by stefanpenner** [comment]
> Example project (including my system and node evnironment info) is here

the project provided seems to produce production builds in a reasonable amount of time:

```
➜  ember-build-prod-test git:(master) time ember b --env production
version: 2.4.1
Built project successfully. Stored in "dist/".
        8.26 real         9.46 user         0.58 sys
```

I suspect not all the required information has been provided.

**Comment by markstuart** [comment]
Apologies, I had added a bunch on info to the README in that project, I'll add it here as well if you prefer?

**Comment by stefanpenner** [comment]
i see:   `// if (!process.env.TINY_MCE) return app.toTree()` sorry didn't notice that before, i can now replicate the issue

**Comment by markstuart** [comment]
Perhaps this issue would make more sense for other readers if I add appropriate context from the example project README as well:

To illustrate this, I have set up two commands: `npm run build:prod` and `npm run build:prod:tinymce`. The second is the same as the first except that it sets a flag to enable the funnelling of TinyMCE in `ember-cli-build.js`.

If you `npm run build:prod` you'll see a lot of console output. When it completes, you can see the time that it took in the final output.

The time difference between `npm run build:prod:tinymce` and `npm run build:prod` is significant on my system.

### My setup:

System info
* Model Name: MacBook Pro
* Processor Name: Intel Core i5
* Processor Speed:  2.4 GHz
* Number of Processors: 1
* Total Number of Cores:  2
* L2 Cache (per Core):  256 KB
* L3 Cache: 3 MB
* Memory: 16 GB

Hard Disk info:
* Device Name:  APPLE SSD SD0256F
* Media Name: APPLE SSD SD0256F Media
* Size: 250.14 GB (250,140,434,432 bytes)
* Medium Type:  SSD

Node info:
`npm -v`: 2.14.21
`node -v`: v4.3.1

### Times:
* `npm run build:prod:tinymce`: ~4.5 min
* `npm run build:prod`: ~15 sec

The main issue seems to happen immediately after this point in the build:
```
  broccoli-merge-trees:TreeMerger (allTrees) build:
 { count: 1, in: '12ms', mergeTime: '0ms', applyPatchTime: '12ms', treeTime: '0ms', entries: 7, indices: 7, overwrite: true } +0ms
```

In both cases the build appears to hang for a period of time at this point, but in the `build:prod:tinymce` version it hangs for a *much* longer time.

**Comment by stefanpenner** [comment]
So the issue is as follows:

uglify is trying to uglify tinymce's already uglified files, and so sad things are happening

Turns out, I have actually provided improved logging for this [here](https://github.com/ef4/broccoli-uglify-sourcemap/pull/12) but it had not yet been merged. (and i forgot about it)

TL;DR files included via `app.import` will end up part of `vendor`, and vendor will be uglified. Uglify uglifying something already uglified is amazingly slow.

Updating you `app.imports` to include the unminified versions
```
  app.import(app.bowerDirectory + '/tinymce/tinymce.js', {destDir: 'assets/tinymce'})
  app.import(app.bowerDirectory + '/tinymce/tinymce.jquery.js', {destDir: 'assets/tinymce'})
```

restore your production builds to something more reasonable:

```
Build successful - 13033ms.

Slowest Trees                                 | Total
----------------------------------------------+---------------------
UglifyWriter                                  | 12254ms

Slowest Trees (cumulative)                    | Total (avg)
----------------------------------------------+---------------------
UglifyWriter (1)                              | 12254ms

```

**Comment by stefanpenner** [comment]
I'm going to mark this as closed, as the logging PR (once merged) would have likely provided a good hint

**Comment by markstuart** [comment]
That is really awesome, thank you so much.

I do have one small question though. I tried adding this to `ember-cli-build.js`:
```js
const tinyMceGlob = '**/tinymce/**'

const app = new EmberApp(defaults, {
  minifyJS: {
   options: {
     exclude: [tinyMceGlob]
   }
  },
  minifyCSS: {
   options: {
     exclude: [tinyMceGlob]
   }
  },
  fingerprint: {
   exclude: [tinyMceGlob]
  }
})
```

Should that have had the same effect? I would have thought that excluding the minified files from ember-cli's minification would have worked in the same way. Or perhaps there is a specific exclude for uglification as well?

**Comment by stefanpenner** [comment]
@markstuart that wont have the same affect, because tinyMCE is already part of vendor.js, so the match would have to be against `vendor.js` which is likely unwanted.

**Comment by markstuart** [comment]
Thank you so much for your time Stefan, love your work :100: :+1:
